### ISSUE TITLE
Type guard with generic `Type[T]

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#1460]
Related: #1421

```python
from typing import Type, Any, TypeVar, Union

T = TypeVar("T")

def g(klass: Type[T], obj: Union[T, int]) -> T:
    assert isinstance(obj, klass)
    return obj  # ERROR, though `obj` has type `klass`
```

Is it possible for Pyright to infer obj is an instance of klass?

**Comment by erictraut** [comment]
Technically, this isn't type _inference_, it's type _narrowing_, but I understand your question.

Yes, the `isinstance` narrowing logic should determine that the type of `obj` is narrowed to type `T` in this case. This will be fixed in the next release.

Here's the [PR](https://github.com/microsoft/pyright/commit/cf222d1ad2a3fe4d2b219d67c91984e85650c913). I used your example above in the test case.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.110, which I just published.
